autowarefoundation/autoware.universe
# feat(behavior_path_planner): debug message for lane change

## Description

Previously only avoidance module publishes debug messages that outputs collision check related stats.

This PR aims to extend debug message to lane change module as well.

Due to the different data structure, it is necessary to consider a proper interface to extract the data in each module while code . Therefore, `visitor design pattern` is used to achieve this.

## Related links

New msg is needed. Please cherry-pick it here. https://github.com/tier4/tier4_autoware_msgs/pull/62

## Tests performed

1. use planning simulator to simulate the following scenario
![Screenshot from 2022-10-25 14-45-23](https://user-images.githubusercontent.com/93502286/197692565-b5335239-f133-49cc-ae82-504efe970592.png)
2. Execute the following command
```
ros2 topic echo /planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/debug/lane_change_debug_message_array
```
3. The output should be as follows
![Screenshot from 2022-10-25 14-47-32](https://user-images.githubusercontent.com/93502286/197692913-60818648-4762-482b-b6ff-7b8c41912550.png)


## Notes for reviewers

Since avoidance module's debug message follows similar interface, the reviewer should also test avoidance module debug message
```
ros2 topic echo /planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/debug/avoidance_debug_message_array
```
![Screenshot from 2022-10-25 14-49-16](https://user-images.githubusercontent.com/93502286/197693131-bbf068fe-6472-422a-a988-442f46926866.png)
![Screenshot from 2022-10-25 14-49-27](https://user-images.githubusercontent.com/93502286/197693137-30c707d7-ab86-460c-87a6-51e665789a75.png)


## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
